sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,104 @@
+import { PollOperation } from "./pollOperation";
+
+type CancelOnProgress = () => void;
+
+export abstract class Poller<TProperties, TResult> {
+  private stopped: boolean;
+  private resolve?: (value: TResult) => void;
+  private reject?: (error: Error) => void;
+  private pollOncePromise?: Promise<void>;
+  private promise: Promise<TResult>;
+  private onProgressCallback?: (op: PollOperation<TProperties>) => void;
+  public operation: PollOperation<TProperties>;
+
+  constructor(operation: PollOperation<TProperties>, stopped: boolean = false) {
+    this.operation = operation;
+    this.stopped = stopped;
+    this.promise = new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
+    });
+    this.operation.state.started = true;
+    if (!this.stopped) {
+      this.startPolling();
+    }
+  }
+
+  abstract async delay(): Promise<void>;
+  abstract async getResult(): Promise<TResult>;
+
+  private async startPolling(): Promise<void> {
+    while (!this.isStopped() && !this.isDone()) {
+      await this.poll();
+      await this.delay();
+    }
+  }
+
+  private async pollOnce(): Promise<void> {
+    try {
+      if (!this.isDone()) {
+        this.operation = await this.operation.update();
+        if (this.isDone() && this.resolve) {
+          this.resolve(await this.getResult());
+        }
+      }
+    } catch (e) {
+      this.operation.state.error = e;
+      if (this.reject) {
+        this.reject(e);
+      }
+    }
+  }
+
+  public poll(): Promise<void> {
+    if (!this.pollOncePromise) {
+      this.pollOncePromise = this.pollOnce();
+      this.pollOncePromise.finally(() => {
+        if (this.onProgressCallback) {
```

Comment:
this one confused me a little, since I thought progress was only when the operation updated in a meaningful way, not after every time polling occurred?

Reply:
Oh! I forgot about the meaningful part. I'll update it!